File: web/src/rendering/geometry.js

import * as Linear     from "./linear"

const vec3 = glMatrix.vec3


export const GLYPH_Cube     = 0
export const GLYPH_Sphere   = 1

export const GLYPH_Box      = 0
export const GLYPH_Cylinder = 1


export function defaultGeometry() {
  return {
    shape: {
      points: []
    , glyph : GLYPH_Cube
    }
  , size : 0
  , color: 0
  , text : ""
  }
}


const MASK_Posn =  1
const MASK_Size =  2
const MASK_Colr =  4
const MASK_Text =  8
const MASK_Glyp = 16


export function deltaPosition(deltaGeometry) {
  return (deltaGeometry.getMask() & MASK_Posn) != 0
}

export function deltaSize(deltaGeometry) {
  return (deltaGeometry.getMask() & MASK_Size) != 0
}

export function deltaColor(deltaGeometry) {
  return (deltaGeometry.getMask() & MASK_Colr) != 0
}

export function deltaText(deltaGeometry) {
  return (deltaGeometry.getMask() & MASK_Text) != 0
}

export function deltaGlyph(deltaGeometry) {
  return (deltaGeometry.getMask() & MASK_Glyp) != 0
}


export const GEOMETRY_Points     = 1
export const GEOMETRY_Polylines  = 2
export const GEOMETRY_Rectangles = 3
export const GEOMETRY_Label      = 4
export const GEOMETRY_Axis       = 5


export function merge(geometry, deltaGeometry) {

  const mask = deltaGeometry.getMask()

  if ((mask & MASK_Posn) != 0) {

    const posx = deltaGeometry.getPosxList()
    const posy = deltaGeometry.getPosyList()
    const posz = deltaGeometry.getPoszList()

    switch (deltaGeometry.getType()) {

      case GEOMETRY_Points:
        {
          const pointses = []
          let i = 0
          deltaGeometry.getCntsList().forEach(function(n) {
            const points = []
            for (let j = 0; j < n; ++j)
              points.push(vec3.fromValues(posx[i], posy[i], posz[i++]))
            pointses.push(points)
          })
          geometry.shape = {
            points: pointses
          , glyph : ("glyph" in geometry.shape) ? geometry.shape.glyph : GLYPH_Cube
          }
          break
        }

      case GEOMETRY_Polylines:
        {
          const polylines = []
          let i = 0
          deltaGeometry.getCntsList().forEach(function(n) {
            const polyline = []
            for (let j = 0; j < n; ++j)
              polyline.push(vec3.fromValues(posx[i], posy[i], posz[i++]))
            polylines.push(polyline)
          })
          geometry.shape = {
            polylines: polylines
//        , glyph    : ("glyph" in geometry.shape) ? geometry.shape.glyph : GLYPH_Cube
          }
          break
        }

      case GEOMETRY_Rectangles:
        {
          const rectangles = []
          let i = 0
          deltaGeometry.getCntsList().forEach(function(n) {
            rectangles.push([
              vec3.fromValues(posx[i], posy[i], posz[i++])
            , vec3.fromValues(posx[i], posy[i], posz[i++])
            , vec3.fromValues(posx[i], posy[i], posz[i++])
            ])
          })
          geometry.shape = {
            rectangles: rectangles
          }
          break
        }

      case GEOMETRY_Label:
        {
          geometry.shape = {
            label: [
              vec3.fromValues(posx[0], posy[0], posz[0])
            , vec3.fromValues(posx[1], posy[1], posz[1])
            , vec3.fromValues(posx[2], posy[2], posz[2])
            ]
          }
          break
        }

      case GEOMETRY_Axis:
        {
          geometry.shape = {
            axis: [
              vec3.fromValues(posx[0], posy[0], posz[0])
            , vec3.fromValues(posx[1], posy[1], posz[1])
            ]
          }
          break
        }

    }

  }

  if ((mask & MASK_Size) != 0)
    geometry.size = deltaGeometry.getSize()

  if ((mask & MASK_Colr) != 0)
    geometry.color = deltaGeometry.getColr()

  if ((mask & MASK_Text) != 0)
    geometry.text = deltaGeometry.getText()

  if ((mask & MASK_Glyp) != 0 && (("points" in geometry.shape) || ("polylines" in geometry.shape)))
    geometry.shape.glyph = deltaGeometry.getGlyp()

}


export function hasAxis(geometry) {
  return "axis" in geometry.shape
}


export function hasLabel(geometry) {
  return "label" in geometry.shape
}


export function hasPoints(geometry) {
  return "points" in geometry.shape
}


export function hasPolylines(geometry) {
  return "polylines" in geometry.shape
}


export function hasRectangles(geometry) {
  return "rectangles" in geometry.shape
}


export function distance(geometry, origin) {

  let dist = Number.MAX_VALUE

  if (hasPoints(geometry)) {

    geometry.shape.points.forEach(pointses => pointses.forEach(point => {
      dist = Math.min(dist, Linear.distancePointSphere(origin, point, geometry.size / 2))
    }))

  } else if (hasPolylines(geometry)) {

    geometry.shape.polylines.map((polyline) => [
      polyline.slice(0, polyline.length - 1)
    , polyline.slice(1, polyline.length    )
    ]).forEach(([u0s, u1s]) =>
      u0s.map(function(u0, i) {
        const u1 = u1s[i]
        dist = Math.min(dist, Linear.distancePointCylinder(origin, u0, u1, geometry.size / 2))
    }))

  } else if (hasRectangles(geometry)) {

    geometry.shape.rectangles.forEach(function([o, u, v]) {
      dist = Math.min(dist, Linear.distancePointRectangle(origin, o, u, v))
    })

  } else if (hasLabel(geometry)) {

  } else if (hasAxis(geometry)) {

    const u0 = geometry.shape.axis[0]
    const u1 = geometry.shape.axis[1]
    dist = Math.min(dist, Linear.distancePointCylinder(origin, u0, u1, geometry.size / 2))

  }

  return dist

}

File: go/infovis/model/Geometry.go
package model


const (
  GLYPH_CUBE   = iota
  GLYPH_SPHERE = iota
)

const (
  GLYPH_BOX      = iota
  GLYPH_CYLINDER = iota
)

const (
  MASK_POSN =  1
  MASK_SIZE =  2
  MASK_COLR =  4
  MASK_TEXT =  8
  MASK_GLYP = 16
)

const (
  GEOMETRY_POINTS     = 1 + iota
  GEOMETRY_POLYLINES  = 1 + iota
  GEOMETRY_RECTANGLES = 1 + iota
  GEOMETRY_LABEL      = 1 + iota
  GEOMETRY_AXIS       = 1 + iota
)


const (
  MESH_CUBE   = iota
  MESH_SPHERE = iota
  MESH_LINE   = iota
  MESH_SQUARE = iota
  MESH_LABEL  = iota
  MESH_ARROW  = iota
)

var MESHES = [...]int32{MESH_CUBE, MESH_SPHERE, MESH_LINE, MESH_SQUARE, MESH_LABEL, MESH_ARROW}


func deltaPosition(geometry *Geometry) bool {
  return geometry.Mask & MASK_POSN != 0
}

func deltaSize(geometry *Geometry) bool {
  return geometry.Mask & MASK_SIZE != 0
}

func deltaColor(geometry *Geometry) bool {
  return geometry.Mask & MASK_COLR != 0
}

func deltaText(geometry *Geometry) bool {
  return geometry.Mask & MASK_TEXT != 0
}

func deltaGlyph(geometry *Geometry) bool {
  return geometry.Mask & MASK_GLYP != 0
}


func merge(geometry *Geometry, delta *Geometry) {
  if deltaPosition(delta) {
    geometry.Cnts = delta.Cnts
    geometry.Posx = delta.Posx
    geometry.Posy = delta.Posy
    geometry.Posz = delta.Posz
  }
  if deltaSize(delta) {
    geometry.Size = delta.Size
  }
  if deltaColor(delta) {
    geometry.Colr = delta.Colr
  }
  if deltaText(delta) {
    geometry.Text = delta.Text
  }
  if deltaGlyph(delta) {
    geometry.Glyp = delta.Glyp
  }
}


/*
function updateDisplay(identifier, geometry, shapeBuffer) {

  const positions = []
  const rotations = []
  const scales    = []
  const color     = geometry.color

  const noRotation = quat.fromValues(0, 0, 0, 1)
  const right = vec3.fromValues(1, 0, 0)
  const back  = vec3.fromValues(0, 0, 1)

  if (Geometry.hasPoints(geometry)) {

    positions.push(...geometry.shape.points.flat())
    rotations.push(...positions.map((point) => noRotation))
    scales.push(...positions.map((point) => vec3.fromValues(geometry.size, geometry.size, geometry.size)))

  } else if (Geometry.hasPolylines(geometry)) {

    geometry.shape.polylines.map((polyline) => [
      polyline.slice(0, polyline.length - 1)
    , polyline.slice(1, polyline.length    )
    ]).forEach(([u0s, u1s]) =>
      u0s.map(function(u0, i) {
        const u1 = u1s[i]
        const ud = vec3.scaleAndAdd(vec3.create(), u1, u0, -1)
        const uc = vec3.scaleAndAdd(vec3.create(), u0, ud, 0.5)
        positions.push(uc)
        rotations.push(Linear.rotationFromVectorPair(right, ud))
        scales.push(vec3.fromValues(vec3.length(ud), geometry.size, geometry.size))
      })
    )

  } else if (Geometry.hasRectangles(geometry)) {

    geometry.shape.rectangles.forEach(function([o, u, v]) {
      const w = vec3.length(vec3.scaleAndAdd(vec3.create(), u, o, -1))
      const h = vec3.length(vec3.scaleAndAdd(vec3.create(), v, o, -1))
      const q = Linear.rotationFromPlane(right, back, o, u, v)
      positions.push(vec3.add(
        vec3.create()
      , o
      , vec3.transformQuat(vec3.create(), vec3.fromValues(w / 2, 0, h / 2), q)
      ))
      rotations.push(q)
      scales.push(vec3.fromValues(w, geometry.size, h))
    })

  } else if (Geometry.hasLabel(geometry)) {

    undefined

  } else if (Geometry.hasAxis(geometry)) {

    const u0 = geometry.shape.axis[0]
    const u1 = geometry.shape.axis[1]
    const ud = vec3.scaleAndAdd(vec3.create(), u1, u0, -1)
    const uc = vec3.scaleAndAdd(vec3.create(), u0, ud, 0.5)

    positions.push(uc)
    rotations.push(Linear.rotationFromVectorPair(right, ud))
    scales.push(vec3.fromValues(vec3.length(ud), geometry.size, geometry.size))

  }

  Buffers.insertPositions(identifier, positions, shapeBuffer)
  Buffers.updateRotations(identifier, rotations, shapeBuffer)
  Buffers.updateScales   (identifier, scales   , shapeBuffer)
  Buffers.updateColor    (identifier, color    , shapeBuffer)

}
*/
